### PR TITLE
Add Max Payout Flag

### PR DESCRIPTION
**Pull request by Sneagan**
### Summary

Add a max payout flag for existing payout tooling.

### Type of Change

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [X] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [X] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [X] Have you squashed all intermediate commits?
- [X] Is there a clear title that explains what the PR does?
- [X] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

**Review by pavelbrm — COMMENTED** [review]
Not ready to approve – unless I am wrong, there is missing part that parses the supplied limit into a decimal, and then that decimal is put in the context. Also, the code can be adjusted so that we don't do the same work again in the loops.

File: tools/settlement/bitflyer/upload.go [hunk 1]
```diff
@@ -248,13 +249,23 @@ func setupSettlementTransactions(
...
 func createBitflyerRequest(
+	ctx context.Context,
 	dryRun *bitflyer.DryRunOption,
 	token string,
 	settlementRequests []settlement.AggregateTransaction,
 ) (*bitflyer.WithdrawToDepositIDBulkPayload, error) {
 	set := []custodian.Transaction{}
 	sourceFrom := ""
+	maxAmountInterface := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey)
+	maxAmount, ok := maxAmountInterface.(decimal.Decimal)
```
[the remark this answers is not in the document]
Comment (Sneagan) [reply tools/settlement/bitflyer/upload.go:260]: Excellent catch.

File: tools/settlement/bitflyer/upload.go [hunk 2]
```diff
@@ -248,13 +249,23 @@ func setupSettlementTransactions(
...
 	dryRun *bitflyer.DryRunOption,
 	token string,
 	settlementRequests []settlement.AggregateTransaction,
 ) (*bitflyer.WithdrawToDepositIDBulkPayload, error) {
 	set := []custodian.Transaction{}
 	sourceFrom := ""
+	maxAmountAny := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey)
+	maxAmount, ok := maxAmountAny.(decimal.Decimal)
+	if !ok {
+		return nil, errors.New("provided max amount is not an integer")
```
Comment (pavelbrm) [inline tools/settlement/bitflyer/upload.go:262]: If there's any intent to make this unit-testable in the future, then replacing this with a pre-defined sentinel error (which can be tested against) would be good.

File: tools/settlement/bitflyer/upload.go [hunk 3]
```diff
@@ -248,13 +249,23 @@ func setupSettlementTransactions(
...
 func createBitflyerRequest(
+	ctx context.Context,
 	dryRun *bitflyer.DryRunOption,
 	token string,
 	settlementRequests []settlement.AggregateTransaction,
 ) (*bitflyer.WithdrawToDepositIDBulkPayload, error) {
 	set := []custodian.Transaction{}
 	sourceFrom := ""
+	maxAmountAny := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey)
+	maxAmount, ok := maxAmountAny.(decimal.Decimal)
```
Comment (pavelbrm) [inline tools/settlement/bitflyer/upload.go:260]: Nit: We could perhaps shrink it to:

```suggestion
	maxAmount, ok := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey).(decimal.Decimal)
```

File: tools/settlement/cmd/uphold.go [hunk 4]
```diff
@@ -231,13 +241,24 @@ func UpholdUpload(
 			Count:   0,
 		}},
 	}
+
+	maxAmountAny := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey)
+	maxAmount, ok := maxAmountAny.(decimal.Decimal)
```
Comment (pavelbrm) [inline tools/settlement/cmd/uphold.go:246]: Similar:

```suggestion
	maxAmount, ok := ctx.Value(appctx.PayoutTxnMaxAmountCTXKey).(decimal.Decimal)
```

File: tools/settlement/cmd/uphold.go [hunk 5]
```diff
@@ -231,13 +241,24 @@ func UpholdUpload(
...
+	if !ok {
+		logger.Panic().Err(err).Msg("provided max amount is not a number")
```
Comment (pavelbrm) [inline tools/settlement/cmd/uphold.go:248]: Nit: Surrounding places to use `Panic`, but the function we are in does return an error. We could return and error and handle it in the caller if we wanted a panic. But also, what value does a panic over error bring here?

File: tools/settlement/cmd/uphold.go [hunk 6]
```diff
@@ -231,13 +241,24 @@ func UpholdUpload(
...
+
 	for i := 0; i < total; i++ {
 		settlementTransaction := &settlementState.Transactions[i]
 
 		if settlementTransaction.IsComplete() || settlementTransaction.IsFailed() {
 			continue
 		}
 
+		if settlementTransaction.Amount.GreaterThan(maxAmount) {
+			logger.Panic().Err(err).Msgf("amount %s exceeds the max %s", settlementTransaction.Amount, maxAmount)
```
Comment (pavelbrm) [inline tools/settlement/cmd/uphold.go:259]: Re: panic – same comment as above.

**Review by pavelbrm — COMMENTED** [review]
Ready to approve as it is (with some comments), but the CI is showing ❌ , which indicates this is not ready to be merged.

Tag me when the tests have been fixed. 😸

**Comment by pavelbrm** [comment]
@Sneagan Can this be closed?

**Pull request closed without merging**